### ISSUE TITLE
(Windows 10) VM crashes with "Illegal instruction" on call to args[0].decode()

### ISSUE THREAD

**Issue by tridius** [hansihe/rustler#101]
I am trying to build a simple NIF library on Windows 10. Right now I am only implementing the default `add` function which is auto-generated when `mix rustler.new` is called. My resulting Rust code looks like this:

```
#[macro_use] extern crate rustler;
#[macro_use] extern crate rustler_codegen;
#[macro_use] extern crate lazy_static;

use rustler::{NifEnv, NifTerm, NifResult, NifEncoder};

mod atoms {
    rustler_atoms! {
        atom ok;
        //atom error;
        //atom __true__ = "true";
        //atom __false__ = "false";
    }
}

rustler_export_nifs! {
    "Elixir.Monolith",
    [("add", 2, add)],
    None
}

fn add<'a>(env: NifEnv<'a>, args: &[NifTerm<'a>]) -> NifResult<NifTerm<'a>> {
    println!("Hello!");
    let num1: i64 = try!(args[0].decode());
    println!("2");
    let num2: i64 = try!(args[1].decode());
    println!("3");

    Ok((atoms::ok(), num1 + num2).encode(env))
}
```
The println!'s are my addition for debugging. When I use `iex -S mix` and try to call `Monolith.add(2,2)`, the "Hello!" is printed, followed by the text "Illegal instruction", then iex crashes completely. This leads me to believe that there is a problem with .decode() for &[NifTerm<'a>] on Windows.

This same project builds and runs perfectly fine on OSX.

The platform on which it's failing has the following characteristics:
Windows 10
rustc 1.17.0 (56124baa9 2017-04-24) (stable)
Elixir 1.4.2
Rustler: {:rustler, github: "hansihe/rustler", sparse: "rustler_mix", branch: "master"}

**Comment by tridius** [comment]
Worth noting: I cloned the repo and tried to run `mix compile` in the "test" directory, and immediately after compiling, the VM also crashes with an "Illegal instruction"

**Comment by hansihe** [comment]
Sorry for the slow response.

Could you try flushing stdout right before the function returns, just to be completely sure the problem actually is in the term decode function?

```
use std::io::{self, Write};
io::stdout().flush().unwrap();
```

**Comment by tridius** [comment]
No problem at all.

I peppered the `add` function with calls to io::stdout().flush().unwrap(), but the function still crashes in the same place.
Also, to be clear, the function does not return, but the entire system seems to crash with no error message other than print the "Illegal instruction" text.

The output with io::stdout().flush().unwrap() between every line in the add function above is this:

```
iex(1)> Monolith.add(2,2)
Hello!
Illegal instruction
```
Additionally, although I have not yet tried this myself, a co-worker tried to use only the .encode() function to encode some ints (essentially just the final line of my add function, with num1 and num2 rust i64s), and the .encode() function also crashed the VM with an "Illegal Instruction" message.

**Comment by hansihe** [comment]
Okey, I reproduced the issue on a windows machine. I know what the issue is now, and I will hopefully have a chance to fix it tomorrow. Thanks for reporting it!

**Comment by tridius** [comment]
That's excellent news! Thanks for such a quick investigation

**Comment by hansihe** [comment]
Sorry for taking so long, I was out of town, and did not have access to the machine I had windows on.

This should be fixed on master by ba9df864b0265fb330dbbc45fe5569aaa5b2e68d.
